### PR TITLE
Configurable Shader and ShaderManager initial setup

### PR DESCRIPTION
**Pull request by matthewjmay**
## Motivation and Context

Initial PR for issue #319.

## How Has This Been Tested

New test cases will be added incrementally for the configurable shaders project as real functionality is implemented. For now, this code isn't being used by existing render pipeline.
All existing test cases pass.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/esp/gfx/Shader.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+              Magnum::Shaders::Phong::Flag::DiffuseTexture,
+          3 /*lights*/);
+    } break;
+
+    default:
+      return nullptr;
+      break;
+  }
+
+  /* Default setup for Phong, shared by all models */
```
Comment (aclegg3) [inline src/esp/gfx/Shader.cpp]: Will all Shaders of same shaderProgram type be collapsed into one setup? (e.g. COLORED_SHADER_PHONG and TEXTURED_SHADER_PHONG different only in config, not separately sub-classed)

File: src/esp/sim/Simulator.h [hunk 2]
```diff
@@ -308,6 +311,8 @@ class Simulator {
 
   gfx::WindowlessContext::uptr context_ = nullptr;
   std::shared_ptr<gfx::Renderer> renderer_ = nullptr;
+  gfx::ShaderManager shaderManager_;
```
Comment (aclegg3) [inline src/esp/sim/Simulator.h]: Stored at this level, we will likely have to pass this down into the `loadScene()` pipeline also to create the correct default shaders for the scene `DrawableGroups` based on `AssetInfo` types. It will also change the C++ test setup pipeline at that point.

File: src/esp/gfx/Shader.cpp (hunk 1 above)
Comment (matthewjmay) [reply src/esp/gfx/Shader.cpp]: Yep, currently we should just have Flat, PrimitiveID, Phong, and PtexMesh ShaderProgram classes, and COLORED_SHADER_PHONG and TEXTURED_SHADER_PHONG will only differ in config

Comment (matthewjmay) [reply src/esp/gfx/Shader.cpp]: Yep, currently we should just have Flat, PrimitiveID, Phong, and PtexMesh ShaderProgram classes, and COLORED_SHADER_PHONG and TEXTURED_SHADER_PHONG will only differ in config

File: src/esp/gfx/Shader.h [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+#include "esp/core/esp.h"
+
+namespace esp {
+namespace gfx {
+
+class Drawable;
+class DrawableGroup;
+class RenderCamera;
+
+// TODO: string type to allow for dynamically added shader types
```
[the remark this answers is not in the document]
Comment (matthewjmay) [reply src/esp/gfx/Shader.h]: I was thinking we were going to allow for dynamically added Shader types to be defined and added in python/javascript, similar to what we are doing with the Sensor factory

File: src/esp/gfx/Shader.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+enum class ShaderType {
```
[the remark this answers is not in the document]
Comment (matthewjmay) [reply src/esp/gfx/Shader.h]: that's exactly right

File: src/esp/gfx/ShaderManager.cpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      std::forward_as_tuple(std::forward<ShaderArgs>(args)...));
+  if (inserted.second) {
+    LOG(INFO) << "Created Shader: " << inserted.first->first;
+    return inserted.first->second;
+  }
+  return nullptr;
+}
+
+bool ShaderManager::deleteShader(const std::string& id) {
+  return shaders_.erase(id);
```
[the remark this answers is not in the document]
Comment (matthewjmay) [reply src/esp/gfx/ShaderManager.cpp]: Shaders will be stored in DrawableGroups as shared pointers, so this won't be a problem. The only repercussion is that if a new shader is made with the same name (which was deleted), DrawableGroups using the old shader won't be updated with the new shader.

File: src/esp/gfx/Shader.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  TEXTURED_SHADER = 4,
+
+  COLORED_SHADER_PHONG = 5,
+  VERTEX_COLORED_SHADER_PHONG = 6,
+  TEXTURED_SHADER_PHONG = 7
+};
+
+struct ShaderConfiguration {
+  ShaderType type = ShaderType::TEXTURED_SHADER;
+  // JSON options;
```
[the remark this answers is not in the document]
Comment (matthewjmay) [reply src/esp/gfx/Shader.h]: there won't be derived configs, instead the factories will all take in JSON options, which will be converted to specific config types based in the specific factory that gets called, before the config is used to construct the ShaderProgram. This is quite flexible since it doesn't require complex inheritance hierarchies, and makes it possible for new configurations to be specified in python/javascript. This is the same approach being taken in the Sensor API redesign.

File: src/esp/sim/Simulator.h (hunk 2 above)
Comment (matthewjmay) [reply src/esp/sim/Simulator.h]: For now, I was planning on passing this into the ResourceManager for use when drawables are created.

File: src/esp/gfx/Shader.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (matthewjmay) [reply src/esp/gfx/Shader.h]: Great idea, we can definitely use that!

I wanted to clarify, these strings won't actually be used in the render loop, instead only used on new shader creation or setting a new config for a created `Shader`. `ShaderType` only exists so that we know which factory function to call when given a `ShaderConfig`. The actual shader lookup based on shader UUID will be done in `ShaderManager` (see #454).

File: src/esp/gfx/Shader.h [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  void setConfiguration(const ShaderConfiguration& config);
+
+  /**
+   * @brief Prepare to draw a @ref DrawableGroup with given @ref RenderCamera
+   *
+   * @return Whether the @ref DrawableGroup can be drawn by the current shader
+   * configuration
+   */
+  bool prepareForDraw(const DrawableGroup& drawables,
+                      const RenderCamera& camera);
```
[the remark this answers is not in the document]
Comment (matthewjmay) [reply src/esp/gfx/Shader.h]: Oops, no idea why I'm actually passing through `DrawableGroup` here. Really this function exists just to set up common shader parameters shared by the whole group, which I don't think can fail. Will update return type to void.

As for the `setConfiguration` idea, any ideas on how to check `isConfigurationCompatible` in a clean way without actually creating a ShaderProgram? I.e. different shader types will have different valid configurations, so checking validity would require creating a new shader via factory (since `Shader` doesn't know about the concrete `ShaderProgram` types), unless I had a separate entry in the lookup table for validation, which I think is overly complicated.

File: src/esp/gfx/ShaderManager.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  /**
+   * @brief Creates a @ref Shader
+   *
+   * @param id    ID of created @ref Shader
+   * @param args  Arguments passed to @ref Shader constructor
+   * @return Pointer to the created @ref Shader, or nullptr if a shader with the
+   * same ID already exists.
+   */
+  template <typename... ShaderArgs>
+  Shader::ptr createShader(std::string id, ShaderArgs&&... args);
```
[the remark this answers is not in the document]
Comment (matthewjmay) [reply src/esp/gfx/ShaderManager.h]: I wanted to be flexible in accepting a shader config here, or copy/move an existing shader into the map.

File: src/esp/gfx/Shader.cpp [hunk 9]
```diff
@@ -20,16 +20,18 @@ namespace gfx {
...
 namespace {
 // TODO: don't hardcode this
-std::unique_ptr<Magnum::GL::AbstractShaderProgram> shaderProgramFactory(
+std::shared_ptr<Magnum::GL::AbstractShaderProgram> shaderProgramFactory(
     ShaderType type,
     const ShaderConfiguration& cfg) {
   switch (type) {
     case ShaderType::INSTANCE_MESH_SHADER: {
-      return std::make_unique<gfx::PrimitiveIDShader>();
+      static auto primitiveShader = std::make_shared<gfx::PrimitiveIDShader>();
```
Comment (mosra) [inline src/esp/gfx/Shader.cpp]: Not sure about the `static` here -- this will mean the instance survives even after the GL context is destroyed, and then when creating a new GL context, this will still return the stale, no longer valid instance.

Same with multithreading -- in case someone would want to run multiple separate simulator instances on separate threads, this makes that impossible.

File: src/esp/gfx/Shader.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (mosra) [reply src/esp/gfx/Shader.h]: I have to admit I was not fully involved in the design discussion for this -- what's the actual definition of a "compatible configuration"? Or how custom can the custom shaders be?

Let's say a custom shader would need an additional buffer texture containing some per-face data to do some fancy shading. Or a PBR shader that additionally needs a metalness texture, or an instanced shader that supplies per-instance matrices via an additional attribute. How would one express that via the configuration? What's the scope of the configuration? Key/value pairs or more? What is a possible use case and what is already out of reach?

I'd say that for a flexible shader management you'd need to:

- have a possibility to call something up-front, before drawing everything that belongs to given shader, for a global setup
- have a possibility to call something before each draw, for a per-mesh setup

In other words, for each new shader, letting users implement their own Drawable that gets used with it, and letting them put things into / operate drawable groups directly. That doesn't need any configuration or configuration validation because users are in charge of that, and ... isn't that kinda already there? I fear this ends up overengineered without really providing the flexibility needed.

[the remark this answers is not in the document]
Comment (matthewjmay) [reply src/esp/gfx/Shader.h]: Thanks for these questions, we definitely need to get the design right before moving further.

The idea is that the "configuration" will specify attributes for shared drawable group setup for any arbitrary type of Shader. A possible use case of this configuration may be to specify type "phong" with attributes numLights, positions, shininess, etc,  or type "flat" with texture flags, vertex colored flags, etc. Now, this `Shader` wrapper class does not know about these underlying shader types, so it must just accept a base type, and the underlying shaderprogram must define what is valid.

So it seems strange that we have the `Shader` wrapper class at all, since we could just allow the client code to configure the underlying `ShaderProgram` directly, before passing it to a DrawableGroup to be used. However, the advantages of this wrapper are:

- The client code doesn't need to know about the underlying shader types being used, then they can just easily pass us a config and we handle shader setup and reuse. Phong implementation could even be swapped out depending on what's registered in the factory.

- The client code doesn't need to store and keep track of their configs (ie. just set and forget, if I want to get it later, and possibly modify it, I can access it with `getConfiguration`

- The configuration can be stored/read in material lighting information in assets (ie. GLtf), instead of always having to be created programmatically.

- The modification/reuse of ShaderPrograms is abstracted. The user shouldn't have to know that adding a light to a phong shader requires me to create a brand new ShaderProgram and set it, but if I just want to move a light I can reuse the existing ShaderProgram. With this wrapper class, they change the configuration in a single uniform way with `setConfiguration`. If necessary, the factories will create a new `ShaderProgram` which supports this new conf. If one already exists (ie. all phong shaders using the same flags and number of lights), we don't need to do anything.

I apologize if any of this isn't clear, its hard to split up the PR's into bite sized chunks that individually make sense, so please let me know if you have questions. Thanks again for the feedback.

File: src/esp/gfx/Shader.cpp (hunk 9 above)
Comment (matthewjmay) [reply src/esp/gfx/Shader.cpp]: Great catch, instead we will have to instantiate the shader factory as a member of the simulator, and inject it into this shader class, to ensure instances are torn down correctly, and are always unique across simulator instances.

Comment (erikwijmans) [reply src/esp/gfx/Shader.cpp]: This can't stay unfortunately.  It will break things immediately.  Tearing down and re-spinning up the simulator is a very common occurrence.

Comment (matthewjmay) [reply src/esp/gfx/Shader.cpp]: removed, I'll implement shader reuse later

**Comment by matthewjmay** [comment]
Just merged in master to allow `DrawableGroup` to actually use the new `Shader` class

**Comment by matthewjmay** [comment]
irrelevant after #489
